lib/DXIL/DxilShaderFlags.cpp

```diff
@@ -562,27 +572,23 @@ ShaderFlags ShaderFlags::CollectShaderFlags(const Function *F,
             }
           }
           break;
-        case DXIL::OpCode::SampleGrad:
         case DXIL::OpCode::SampleLevel:
         case DXIL::OpCode::SampleCmpLevelZero:
-          if (!isa<Constant>(CI->getArgOperand(
-                  DXIL::OperandIndex::kTextureSampleOffset0OpIdx)) ||
-              !isa<Constant>(CI->getArgOperand(
-                  DXIL::OperandIndex::kTextureSampleOffset1OpIdx)) ||
-              !isa<Constant>(CI->getArgOperand(
-                  DXIL::OperandIndex::kTextureSampleOffset2OpIdx)))
-            hasAdvancedTextureOps = true;
+          hasAdvancedTextureOps |= hasNonConstantSampleOffsets(CI);
+          break;
+        case DXIL::OpCode::SampleGrad:
+        case DXIL::OpCode::SampleCmpGrad:
+          hasAdvancedTextureOps |= hasNonConstantSampleOffsets(CI);
+          if (!isa<UndefValue>(CI->getArgOperand(CI->getNumArgOperands() - 1)))
+            hasTiledResources = true;
           break;
         case DXIL::OpCode::Sample:
         case DXIL::OpCode::SampleBias:
         case DXIL::OpCode::SampleCmp:
-          if (!isa<Constant>(CI->getArgOperand(
-                  DXIL::OperandIndex::kTextureSampleOffset0OpIdx)) ||
-              !isa<Constant>(CI->getArgOperand(
-                  DXIL::OperandIndex::kTextureSampleOffset1OpIdx)) ||
-              !isa<Constant>(CI->getArgOperand(
-                  DXIL::OperandIndex::kTextureSampleOffset2OpIdx)))
-            hasAdvancedTextureOps = true;
+        case DXIL::OpCode::SampleCmpBias:
+          hasAdvancedTextureOps |= hasNonConstantSampleOffsets(CI);
+          if (!isa<UndefValue>(CI->getArgOperand(CI->getNumArgOperands() - 1)))
```

Comment:
I think this should consider a constant value of `0.0f` as not using the clamp as well.